clang/lib/Sema/SemaCoroutine.cpp

```diff
@@ -418,39 +448,60 @@ static ReadySuspendResumeResult buildCoawaitCalls(Sema &S, VarDecl *CoroPromise,
     return Calls;
   }
   Expr *CoroHandle = CoroHandleRes.get();
-  CallExpr *AwaitSuspend = cast_or_null<CallExpr>(
-      BuildSubExpr(ACT::ACT_Suspend, "await_suspend", CoroHandle));
+  auto *AwaitSuspend = [&]() -> CallExpr * {
+    auto *SubExpr = BuildSubExpr(ACT::ACT_Suspend, "await_suspend", CoroHandle);
+    if (!SubExpr)
+      return nullptr;
+    if (auto *E = dyn_cast<CXXBindTemporaryExpr>(SubExpr)) {
+      // This happens when await_suspend return type is not trivially
+      // destructible. This doesn't happen for the permitted return types of
+      // such function. Diagnose it later.
+      return cast_or_null<CallExpr>(E->getSubExpr());
+    } else {
+      return cast_or_null<CallExpr>(SubExpr);
+    }
```

Comment:
This is necessary for the diagnostic to work. If the return type of `await_suspend` is a type that has a user defined destructor we will get a `CXXBindTemporaryExpr`. The original `cast_or_null` here was the reason it was crashing.

Reply:
I mean, how about diagnose `InvalidAwaitSuspendReturnType` directly if it is null. And we can use `dyn_cast_if_present` instead of `cast_or_null`. WDYT?

Reply:
I was hoping to obtain the return type from the CallExpr.

Reply:
I am still concerning the change. What's the type of `SubExpr`? Can we get the return type of SubExpr directly instead of accessing it by CallExpr?

Reply:
I feel that that's harder than finding the `CallExpr` and then `getReturnType`. What's your suggestion?

Reply:
Currently the diagnostic will read:

```
error: return type of 'await_suspend' is required to be 'void', 'bool', or 'std::coroutine_handle' (have 'Status')
```

With what you suggested, I cannot print the "have 'Status'" part any more.

Reply:
Attempted to drop the `CallExpr` dependence further. I can't diagnose at the location of the `await_suspend()` anymore without a `CallExpr`.

Reply:
Thanks. My motivation is that the small pattern match style looks really not good. And if it is possible to get the RecordDecl by the type of `Operand` and find the `await_suspend` member function? If yes, we can still get the location of `await_suspend`.

And if we have problems printing `(have 'Status')` part, I feel it may not be a big problem since the underlying issue is clear.

Reply:
Not sure what `Operand` means here. Any code pointers?

Reply:
Sorry. The `Operand` is referred to the definitions in line 405, whose type should be the type of the corresponding `awaiter`. The idea here is, if we can get the type of the `awaiter` directly, we can avoid the pattern match here.

Reply:
I still want to fix this. Let me rebase and revive it.